### PR TITLE
Fix Consumer not getting messages after filter update

### PR DESCRIPTION
**Pull request by Jarema**
Found during work on multiple filters

After consumer filter update:
1. Wildcard bool was not updated
2. Sublist for consumers was not updated

Signed-off-by: Tomasz Pietrek <tomasz@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/consumer.go [hunk 1]
```diff
@@ -1591,6 +1591,22 @@ func (o *consumer) updateConfig(cfg *ConsumerConfig) error {
 		}
 	}
 
+	if o.cfg.FilterSubject != cfg.FilterSubject {
+		if cfg.FilterSubject != _EMPTY_ && subjectHasWildcard(cfg.FilterSubject) {
+			o.filterWC = true
+		}
+		subject := cfg.FilterSubject
```
Comment (derekcollison) [inline server/consumer.go:1598]: Could move this to top and use subject in comparison above.

File: server/consumer.go [hunk 2]
```diff
@@ -1591,6 +1591,22 @@ func (o *consumer) updateConfig(cfg *ConsumerConfig) error {
 		}
 	}
 
+	if o.cfg.FilterSubject != cfg.FilterSubject {
+		if cfg.FilterSubject != _EMPTY_ && subjectHasWildcard(cfg.FilterSubject) {
```
Comment (derekcollison) [inline server/consumer.go:1595]: Could just say..

```go
if cfg.FilterSubject != _EMPTY_ {
  o.filterWC = subjectHasWildcard(cfg.FilterSubject)
}
```

File: server/consumer.go [hunk 3]
```diff
@@ -1591,6 +1591,22 @@ func (o *consumer) updateConfig(cfg *ConsumerConfig) error {
...
+		subject := cfg.FilterSubject
+		if subject == _EMPTY_ {
+			subject = fwcs
+		}
+		sub := &subscription{subject: []byte(subject), icb: o.processStreamSignal}
+		o.mset.clsMu.Lock()
+		o.mset.csl.Remove(o.sigSub)
+		o.mset.csl.Insert(sub)
+		o.sigSub = sub
+		o.mset.clsMu.Unlock()
```
Comment (derekcollison) [inline server/consumer.go:1607]: mset does some accounting here we need to track.

See func (mset *stream) removeConsumer(o *consumer)

So might want to make stream func

```go
func (mset *Stream) swapSigSubs(o *consumer)
```

Also, make sure o.sigSub != nil before calling remove I think.

**Review by derekcollison — CHANGES_REQUESTED** [review]

**Comment by Jarema** [comment]
@derekcollison, all review comments addressed.
Thanks for pointing out the missing accounting!

File: server/jetstream_test.go [hunk 4]
```diff
@@ -19294,3 +19294,72 @@ func TestJetStreamMsgBlkFailOnKernelFault(t *testing.T) {
 			friendlyBytes(si.Config.MaxBytes), friendlyBytes(int64(si.State.Bytes)))
 	}
 }
+
+func TestJetStreamConsumerFilterUpdate(t *testing.T) {
+
```
Comment (derekcollison) [inline server/jetstream_test.go:19299]: Remove newline.

**Review by derekcollison — APPROVED** [review]
LGTM - Small nit

**Pull request merged**